pipes/WDL/workflows/merge_bams_bulk.wdl

```diff
@@ -0,0 +1,69 @@
+import "tasks_demux.wdl" as demux
+
+workflow merge_bams_bulk {
+    Array[File]+ in_bams # any order
+    File in_bam_out_bam_table # first column: input bam file basename, second column: output bam file basename (one line per INPUT file)
+    File? reheader_table
+    String? docker="quay.io/broadinstitute/viral-core"
+    
+    # removes ".bam"s from ends of filenames in in_bam_out_bam_table
+    call clean_in_bam_out_bam_table {
+        input: table = in_bam_out_bam_table
+    }
+    
+    # generates map with key: input bam file name -> value: output bam file basename
+    Map[String, String] in_bam_to_out_bam = read_map(clean_in_bam_out_bam_table.clean_table)
+    
+    # retrieves unique output bam file basenames (no repeats)
+    call unique_values_in_second_column {
+        input: table = clean_in_bam_out_bam_table.clean_table
+    }
+    Array[String] out_bams = unique_values_in_second_column.unique_values
+    
+    # collects and merges input bam files for each output bam file
+    scatter (out_bam in out_bams) {
+        String placeholder = write_map(in_bam_to_out_bam) # need to touch map in outer scatter for it to be seen in inner scatter
+        
+        # retrieves the input bam files for this output bam file
+        scatter (in_bam in in_bams) {
+            String in_bam_basename = basename(in_bam, ".bam")
+            if(in_bam_to_out_bam[in_bam_basename] == out_bam) {
+                File relevant_in_bam = in_bam
+            }
+        }
+        Array[File] relevant_in_bams = select_all(relevant_in_bam) # gathers input bam files from the scatter
+        
+        # merges the relevant input bam files to produce this output file
+        call demux.merge_and_reheader_bams {
+            input:
+                out_basename = basename(out_bam, ".bam"),
+                in_bams = relevant_in_bams,
+                reheader_table = reheader_table,
+                docker = docker
+        }
+    }
+}
+
+task clean_in_bam_out_bam_table {
+    File table
+    
+    command {
+    	cat ${table} | sed 's/[.]bam$//g' | sed $'s/[.]bam\t/\t/g' | tee in_bam_out_bam_table
+    }
+    
+    output {
+        File clean_table = "in_bam_out_bam_table"
+    }
+}
```

Comment:
This task seems unnecessary -- you're already calling the WDL `basename` function on all the members of the table, so why do it twice (once in WDL, once in a task)?

Reply:
(As you probably already know,) my thinking here was to deal with the user potentially including the .bam extension in the input and/or output bam filenames listed in in_bam_out_bam_table. I'd like to be able to handle, for example, a mapping like

```
input_a1.bam	output_a
input_a2	output_a
input_b1	output_b.bam
input_b2	output_b
input_c1.bam	output_c.bam
input_d1	output_d.bam
```
or even a more reasonable

```
input_a1.bam	output_a.bam
input_a2.bam	output_a.bam
```

vs.

```
input_a1	output_a
input_a2	output_a
```

I ran into problems with this line:
`if(in_bam_to_out_bam[in_bam_basename] == out_bam)`

Note that `in_bam_basename` is the basename of an input bam file from the list of bam files provided by the user (`Array[File]+ in_bams`), while `in_bam_to_out_bam` is the above mapping file provided by the user. (This map is built directly from the contents of the file, and is not mutable.)

This would be fine if we could check the map for `in_bam_basename` and `in_bam_basename + ".bam"`. I couldn't find a way to check if a key is present in a map. I couldn't figure out how to do that, but if the key is not in the map, `in_bam_to_out_bam[in_bam_basename]` gives me an error. So I decided to clean out the `.bam`s from inside the user-provided mapping file before the map is created.